### PR TITLE
Avoid data race by calling forceCompState only on commandRuntime.Run

### LINKED ISSUE #3512: Possible data race when commandRuntime fails to write connection information
If the command runtime fails to write the connection information to a component, it sets the component state to failed. However it happen on a diffferent goroutine than the runtime compoenent run loop, which also sets the component state, thus causing a data race

For confirmed bugs, please report:
- Version: 8.10+
- Operating System: all
- Discuss Forum URL: N/A
- Steps to Reproduce: 
    - run the test `pkg/component/runtime TestManager_FakeInput_RestartsOnMissedCheckins`
    - make [comm.WriteConnInfo(info.Stdin)
](https://github.com/elastic/elastic-agent/blob/main/pkg/component/runtime/command.go#L423) to return an error.

The issue was found on [this](https://fleet-ci.elastic.co/blue/organizations/jenkins/elastic-agent%2Felastic-agent-mbp/detail/PR-3207/29/pipeline/612/) CI build

The logs as the build will eventually be deleted:

```
[2023-10-04T08:29:45.548Z] === FAIL: pkg/component/runtime TestManager_FakeInput_RestartsOnMissedCheckins (2.61s)

[2023-10-04T08:29:45.548Z]     manager_test.go:2476: component state changed: {State:STARTING Message:Starting: spawned pid '3936' Units:map[{UnitType:input UnitID:fake-input}:{State:STARTING Message:Starting: spawned pid '3936' Payload:map[] unitState:0 unitMessage:Starting unitPayload:map[] configStateIdx:0 err:<nil>}] Features:<nil> FeaturesIdx:0 Component:<nil> ComponentIdx:0 VersionInfo:{Name: Version: Meta:map[]} expectedUnits:map[{UnitType:0 UnitID:fake-input}:{state:2 configStateIdx:1 config:0xc0002b2000 err:<nil> logLevel:0}] expectedFeatures:<nil> expectedFeaturesIdx:1 expectedComponent:<nil> expectedComponentIdx:1}

[2023-10-04T08:29:45.548Z]     manager_test.go:2476: component state changed: {State:DEGRADED Message:Degraded: pid '3936' missed 1 check-in Units:map[{UnitType:input UnitID:fake-input}:{State:STARTING Message:Starting: spawned pid '3936' Payload:map[] unitState:0 unitMessage:Starting unitPayload:map[] configStateIdx:0 err:<nil>}] Features:<nil> FeaturesIdx:0 Component:<nil> ComponentIdx:0 VersionInfo:{Name: Version: Meta:map[]} expectedUnits:map[{UnitType:0 UnitID:fake-input}:{state:2 configStateIdx:1 config:0xc0002b2000 err:<nil> logLevel:0}] expectedFeatures:<nil> expectedFeaturesIdx:1 expectedComponent:<nil> expectedComponentIdx:1}

[2023-10-04T08:29:45.548Z]     manager_test.go:2476: component state changed: {State:DEGRADED Message:Degraded: pid '3936' missed 2 check-ins Units:map[{UnitType:input UnitID:fake-input}:{State:STARTING Message:Starting: spawned pid '3936' Payload:map[] unitState:0 unitMessage:Starting unitPayload:map[] configStateIdx:0 err:<nil>}] Features:<nil> FeaturesIdx:0 Component:<nil> ComponentIdx:0 VersionInfo:{Name: Version: Meta:map[]} expectedUnits:map[{UnitType:0 UnitID:fake-input}:{state:2 configStateIdx:1 config:0xc0002b2000 err:<nil> logLevel:0}] expectedFeatures:<nil> expectedFeaturesIdx:1 expectedComponent:<nil> expectedComponentIdx:1}

[2023-10-04T08:29:45.548Z]     manager_test.go:2476: component state changed: {State:FAILED Message:Failed: pid '3936' missed 3 check-ins and will be killed Units:map[{UnitType:input UnitID:fake-input}:{State:FAILED Message:Failed: pid '3936' missed 3 check-ins and will be killed Payload:map[] unitState:0 unitMessage:Starting unitPayload:map[] configStateIdx:0 err:<nil>}] Features:<nil> FeaturesIdx:0 Component:<nil> ComponentIdx:0 VersionInfo:{Name: Version: Meta:map[]} expectedUnits:map[{UnitType:0 UnitID:fake-input}:{state:2 configStateIdx:1 config:0xc0002b2000 err:<nil> logLevel:0}] expectedFeatures:<nil> expectedFeaturesIdx:1 expectedComponent:<nil> expectedComponentIdx:1}

[2023-10-04T08:29:45.548Z] ==================

[2023-10-04T08:29:45.548Z] WARNING: DATA RACE

[2023-10-04T08:29:45.548Z] Read at 0x00c0003c0750 by goroutine 75:

[2023-10-04T08:29:45.548Z]   github.com/elastic/elastic-agent/pkg/component/runtime.(*ComponentState).forceState()

[2023-10-04T08:29:45.548Z]       C:/Users/jenkins/workspace/_agent_elastic-agent-mbp_PR-3207/src/github.com/elastic/elastic-agent/pkg/component/runtime/state.go:468 +0x64

[2023-10-04T08:29:45.548Z]   github.com/elastic/elastic-agent/pkg/component/runtime.(*commandRuntime).forceCompState()

[2023-10-04T08:29:45.548Z]       C:/Users/jenkins/workspace/_agent_elastic-agent-mbp_PR-3207/src/github.com/elastic/elastic-agent/pkg/component/runtime/command.go:311 +0x8c

[2023-10-04T08:29:45.548Z]   github.com/elastic/elastic-agent/pkg/component/runtime.(*commandRuntime).startWatcher.func1()

[2023-10-04T08:29:45.549Z]       C:/Users/jenkins/workspace/_agent_elastic-agent-mbp_PR-3207/src/github.com/elastic/elastic-agent/pkg/component/runtime/command.go:425 +0x184

[2023-10-04T08:29:45.549Z] 

[2023-10-04T08:29:45.549Z] Previous write at 0x00c0003c0750 by goroutine 74:

[2023-10-04T08:29:45.549Z]   github.com/elastic/elastic-agent/pkg/component/runtime.(*ComponentState).forceState()

[2023-10-04T08:29:45.549Z]       C:/Users/jenkins/workspace/_agent_elastic-agent-mbp_PR-3207/src/github.com/elastic/elastic-agent/pkg/component/runtime/state.go:469 +0xd9

[2023-10-04T08:29:45.549Z]   github.com/elastic/elastic-agent/pkg/component/runtime.(*commandRuntime).forceCompState()

[2023-10-04T08:29:45.549Z]       C:/Users/jenkins/workspace/_agent_elastic-agent-mbp_PR-3207/src/github.com/elastic/elastic-agent/pkg/component/runtime/command.go:311 +0x8c

[2023-10-04T08:29:45.549Z]   github.com/elastic/elastic-agent/pkg/component/runtime.(*commandRuntime).Run()

[2023-10-04T08:29:45.549Z]       C:/Users/jenkins/workspace/_agent_elastic-agent-mbp_PR-3207/src/github.com/elastic/elastic-agent/pkg/component/runtime/command.go:160 +0x6cf

[2023-10-04T08:29:45.549Z]   github.com/elastic/elastic-agent/pkg/component/runtime.(*componentRuntimeState).runLoop.func1()

[2023-10-04T08:29:45.549Z]       C:/Users/jenkins/workspace/_agent_elastic-agent-mbp_PR-3207/src/github.com/elastic/elastic-agent/pkg/component/runtime/runtime.go:137 +0x11c

[2023-10-04T08:29:45.549Z]   github.com/elastic/elastic-agent/internal/pkg/runner.Start.func1()

[2023-10-04T08:29:45.549Z]       C:/Users/jenkins/workspace/_agent_elastic-agent-mbp_PR-3207/src/github.com/elastic/elastic-agent/internal/pkg/runner/runner.go:69 +0x62

[2023-10-04T08:29:45.549Z] 

[2023-10-04T08:29:45.549Z] Goroutine 75 (running) created at:

[2023-10-04T08:29:45.549Z]   github.com/elastic/elastic-agent/pkg/component/runtime.(*commandRuntime).startWatcher()

[2023-10-04T08:29:45.549Z]       C:/Users/jenkins/workspace/_agent_elastic-agent-mbp_PR-3207/src/github.com/elastic/elastic-agent/pkg/component/runtime/command.go:422 +0x11c

[2023-10-04T08:29:45.549Z]   github.com/elastic/elastic-agent/pkg/component/runtime.(*commandRuntime).start()

[2023-10-04T08:29:45.549Z]       C:/Users/jenkins/workspace/_agent_elastic-agent-mbp_PR-3207/src/github.com/elastic/elastic-agent/pkg/component/runtime/command.go:387 +0x12de

[2023-10-04T08:29:45.549Z]   github.com/elastic/elastic-agent/pkg/component/runtime.(*commandRuntime).Run()

[2023-10-04T08:29:45.549Z]       C:/Users/jenkins/workspace/_agent_elastic-agent-mbp_PR-3207/src/github.com/elastic/elastic-agent/pkg/component/runtime/command.go:154 +0x53e

[2023-10-04T08:29:45.549Z]   github.com/elastic/elastic-agent/pkg/component/runtime.(*componentRuntimeState).runLoop.func1()

[2023-10-04T08:29:45.549Z]       C:/Users/jenkins/workspace/_agent_elastic-agent-mbp_PR-3207/src/github.com/elastic/elastic-agent/pkg/component/runtime/runtime.go:137 +0x11c

[2023-10-04T08:29:45.549Z]   github.com/elastic/elastic-agent/internal/pkg/runner.Start.func1()

[2023-10-04T08:29:45.549Z]       C:/Users/jenkins/workspace/_agent_elastic-agent-mbp_PR-3207/src/github.com/elastic/elastic-agent/internal/pkg/runner/runner.go:69 +0x62

[2023-10-04T08:29:45.549Z] 

[2023-10-04T08:29:45.549Z] Goroutine 74 (running) created at:

[2023-10-04T08:29:45.549Z]   github.com/elastic/elastic-agent/internal/pkg/runner.Start()

[2023-10-04T08:29:45.549Z]       C:/Users/jenkins/workspace/_agent_elastic-agent-mbp_PR-3207/src/github.com/elastic/elastic-agent/internal/pkg/runner/runner.go:68 +0x296

[2023-10-04T08:29:45.549Z]   github.com/elastic/elastic-agent/pkg/component/runtime.(*componentRuntimeState).runLoop()

[2023-10-04T08:29:45.549Z]       C:/Users/jenkins/workspace/_agent_elastic-agent-mbp_PR-3207/src/github.com/elastic/elastic-agent/pkg/component/runtime/runtime.go:135 +0xe8

[2023-10-04T08:29:45.549Z]   github.com/elastic/elastic-agent/pkg/component/runtime.newComponentRuntimeState.func1()

[2023-10-04T08:29:45.549Z]       C:/Users/jenkins/workspace/_agent_elastic-agent-mbp_PR-3207/src/github.com/elastic/elastic-agent/pkg/component/runtime/runtime.go:128 +0x39

[2023-10-04T08:29:45.549Z] ==================

[2023-10-04T08:29:45.549Z] ==================

[2023-10-04T08:29:45.549Z] WARNING: DATA RACE

[2023-10-04T08:29:45.549Z] Read at 0x00c0003c0758 by goroutine 75:

[2023-10-04T08:29:45.549Z]   github.com/elastic/elastic-agent/pkg/component/runtime.(*ComponentState).forceState()

[2023-10-04T08:29:45.549Z]       C:/Users/jenkins/workspace/_agent_elastic-agent-mbp_PR-3207/src/github.com/elastic/elastic-agent/pkg/component/runtime/state.go:468 +0x84

[2023-10-04T08:29:45.549Z]   github.com/elastic/elastic-agent/pkg/component/runtime.(*commandRuntime).forceCompState()

[2023-10-04T08:29:45.549Z]       C:/Users/jenkins/workspace/_agent_elastic-agent-mbp_PR-3207/src/github.com/elastic/elastic-agent/pkg/component/runtime/command.go:311 +0x8c

[2023-10-04T08:29:45.549Z]   github.com/elastic/elastic-agent/pkg/component/runtime.(*commandRuntime).startWatcher.func1()

[2023-10-04T08:29:45.549Z]       C:/Users/jenkins/workspace/_agent_elastic-agent-mbp_PR-3207/src/github.com/elastic/elastic-agent/pkg/component/runtime/command.go:425 +0x184

[2023-10-04T08:29:45.549Z] 

[2023-10-04T08:29:45.549Z] Previous write at 0x00c0003c0758 by goroutine 74:

[2023-10-04T08:29:45.549Z]   github.com/elastic/elastic-agent/pkg/component/runtime.(*ComponentState).forceState()

[2023-10-04T08:29:45.549Z]       C:/Users/jenkins/workspace/_agent_elastic-agent-mbp_PR-3207/src/github.com/elastic/elastic-agent/pkg/component/runtime/state.go:470 +0xfb

[2023-10-04T08:29:45.549Z]   github.com/elastic/elastic-agent/pkg/component/runtime.(*commandRuntime).forceCompState()

[2023-10-04T08:29:45.549Z]       C:/Users/jenkins/workspace/_agent_elastic-agent-mbp_PR-3207/src/github.com/elastic/elastic-agent/pkg/component/runtime/command.go:311 +0x8c

[2023-10-04T08:29:45.549Z]   github.com/elastic/elastic-agent/pkg/component/runtime.(*commandRuntime).Run()

[2023-10-04T08:29:45.549Z]       C:/Users/jenkins/workspace/_agent_elastic-agent-mbp_PR-3207/src/github.com/elastic/elastic-agent/pkg/component/runtime/command.go:160 +0x6cf

[2023-10-04T08:29:45.549Z]   github.com/elastic/elastic-agent/pkg/component/runtime.(*componentRuntimeState).runLoop.func1()

[2023-10-04T08:29:45.549Z]       C:/Users/jenkins/workspace/_agent_elastic-agent-mbp_PR-3207/src/github.com/elastic/elastic-agent/pkg/component/runtime/runtime.go:137 +0x11c

[2023-10-04T08:29:45.549Z]   github.com/elastic/elastic-agent/internal/pkg/runner.Start.func1()

[2023-10-04T08:29:45.549Z]       C:/Users/jenkins/workspace/_agent_elastic-agent-mbp_PR-3207/src/github.com/elastic/elastic-agent/internal/pkg/runner/runner.go:69 +0x62

[2023-10-04T08:29:45.549Z] 

[2023-10-04T08:29:45.549Z] Goroutine 75 (running) created at:

[2023-10-04T08:29:45.549Z]   github.com/elastic/elastic-agent/pkg/component/runtime.(*commandRuntime).startWatcher()

[2023-10-04T08:29:45.549Z]       C:/Users/jenkins/workspace/_agent_elastic-agent-mbp_PR-3207/src/github.com/elastic/elastic-agent/pkg/component/runtime/command.go:422 +0x11c

[2023-10-04T08:29:45.549Z]   github.com/elastic/elastic-agent/pkg/component/runtime.(*commandRuntime).start()

[2023-10-04T08:29:45.549Z]       C:/Users/jenkins/workspace/_agent_elastic-agent-mbp_PR-3207/src/github.com/elastic/elastic-agent/pkg/component/runtime/command.go:387 +0x12de

[2023-10-04T08:29:45.549Z]   github.com/elastic/elastic-agent/pkg/component/runtime.(*commandRuntime).Run()

[2023-10-04T08:29:45.549Z]       C:/Users/jenkins/workspace/_agent_elastic-agent-mbp_PR-3207/src/github.com/elastic/elastic-agent/pkg/component/runtime/command.go:154 +0x53e

[2023-10-04T08:29:45.549Z]   github.com/elastic/elastic-agent/pkg/component/runtime.(*componentRuntimeState).runLoop.func1()

[2023-10-04T08:29:45.549Z]       C:/Users/jenkins/workspace/_agent_elastic-agent-mbp_PR-3207/src/github.com/elastic/elastic-agent/pkg/component/runtime/runtime.go:137 +0x11c

[2023-10-04T08:29:45.549Z]   github.com/elastic/elastic-agent/internal/pkg/runner.Start.func1()

[2023-10-04T08:29:45.550Z]       C:/Users/jenkins/workspace/_agent_elastic-agent-mbp_PR-3207/src/github.com/elastic/elastic-agent/internal/pkg/runner/runner.go:69 +0x62

[2023-10-04T08:29:45.550Z] 

[2023-10-04T08:29:45.550Z] Goroutine 74 (running) created at:

[2023-10-04T08:29:45.550Z]   github.com/elastic/elastic-agent/internal/pkg/runner.Start()

[2023-10-04T08:29:45.550Z]       C:/Users/jenkins/workspace/_agent_elastic-agent-mbp_PR-3207/src/github.com/elastic/elastic-agent/internal/pkg/runner/runner.go:68 +0x296

[2023-10-04T08:29:45.550Z]   github.com/elastic/elastic-agent/pkg/component/runtime.(*componentRuntimeState).runLoop()

[2023-10-04T08:29:45.550Z]       C:/Users/jenkins/workspace/_agent_elastic-agent-mbp_PR-3207/src/github.com/elastic/elastic-agent/pkg/component/runtime/runtime.go:135 +0xe8

[2023-10-04T08:29:45.550Z]   github.com/elastic/elastic-agent/pkg/component/runtime.newComponentRuntimeState.func1()

[2023-10-04T08:29:45.550Z]       C:/Users/jenkins/workspace/_agent_elastic-agent-mbp_PR-3207/src/github.com/elastic/elastic-agent/pkg/component/runtime/runtime.go:128 +0x39

[2023-10-04T08:29:45.550Z] ==================

[2023-10-04T08:29:45.550Z] ==================

[2023-10-04T08:29:45.550Z] WARNING: DATA RACE

[2023-10-04T08:29:45.550Z] Read at 0x00c000dcf3e0 by goroutine 75:

[2023-10-04T08:29:45.550Z]   runtime.mapiterinit()

[2023-10-04T08:29:45.550Z]       C:/Users/jenkins/workspace/_agent_elastic-agent-mbp_PR-3207/.gvm/versions/go1.20.8.windows.amd64/src/runtime/map.go:815 +0x0

[2023-10-04T08:29:45.550Z]   github.com/elastic/elastic-agent/pkg/component/runtime.(*ComponentState).forceState()

[2023-10-04T08:29:45.550Z]       C:/Users/jenkins/workspace/_agent_elastic-agent-mbp_PR-3207/src/github.com/elastic/elastic-agent/pkg/component/runtime/state.go:473 +0x1a6

[2023-10-04T08:29:45.550Z]   github.com/elastic/elastic-agent/pkg/component/runtime.(*commandRuntime).forceCompState()

[2023-10-04T08:29:45.550Z]       C:/Users/jenkins/workspace/_agent_elastic-agent-mbp_PR-3207/src/github.com/elastic/elastic-agent/pkg/component/runtime/command.go:311 +0x8c

[2023-10-04T08:29:45.550Z]   github.com/elastic/elastic-agent/pkg/component/runtime.(*commandRuntime).startWatcher.func1()

[2023-10-04T08:29:45.550Z]       C:/Users/jenkins/workspace/_agent_elastic-agent-mbp_PR-3207/src/github.com/elastic/elastic-agent/pkg/component/runtime/command.go:425 +0x184

[2023-10-04T08:29:45.550Z] 

[2023-10-04T08:29:45.550Z] Previous write at 0x00c000dcf3e0 by goroutine 74:

[2023-10-04T08:29:45.550Z]   runtime.mapassign()

[2023-10-04T08:29:45.550Z]       C:/Users/jenkins/workspace/_agent_elastic-agent-mbp_PR-3207/.gvm/versions/go1.20.8.windows.amd64/src/runtime/map.go:578 +0x0

[2023-10-04T08:29:45.550Z]   github.com/elastic/elastic-agent/pkg/component/runtime.(*ComponentState).forceState()

[2023-10-04T08:29:45.550Z]       C:/Users/jenkins/workspace/_agent_elastic-agent-mbp_PR-3207/src/github.com/elastic/elastic-agent/pkg/component/runtime/state.go:489 +0x394

[2023-10-04T08:29:45.550Z]   github.com/elastic/elastic-agent/pkg/component/runtime.(*commandRuntime).forceCompState()

[2023-10-04T08:29:45.550Z]       C:/Users/jenkins/workspace/_agent_elastic-agent-mbp_PR-3207/src/github.com/elastic/elastic-agent/pkg/component/runtime/command.go:311 +0x8c

[2023-10-04T08:29:45.551Z]   github.com/elastic/elastic-agent/pkg/component/runtime.(*commandRuntime).Run()

[2023-10-04T08:29:45.551Z]       C:/Users/jenkins/workspace/_agent_elastic-agent-mbp_PR-3207/src/github.com/elastic/elastic-agent/pkg/component/runtime/command.go:160 +0x6cf

[2023-10-04T08:29:45.551Z]   github.com/elastic/elastic-agent/pkg/component/runtime.(*componentRuntimeState).runLoop.func1()

[2023-10-04T08:29:45.551Z]       C:/Users/jenkins/workspace/_agent_elastic-agent-mbp_PR-3207/src/github.com/elastic/elastic-agent/pkg/component/runtime/runtime.go:137 +0x11c

[2023-10-04T08:29:45.551Z]   github.com/elastic/elastic-agent/internal/pkg/runner.Start.func1()

[2023-10-04T08:29:45.551Z]       C:/Users/jenkins/workspace/_agent_elastic-agent-mbp_PR-3207/src/github.com/elastic/elastic-agent/internal/pkg/runner/runner.go:69 +0x62

[2023-10-04T08:29:45.551Z] 

[2023-10-04T08:29:45.551Z] Goroutine 75 (running) created at:

[2023-10-04T08:29:45.551Z]   github.com/elastic/elastic-agent/pkg/component/runtime.(*commandRuntime).startWatcher()

[2023-10-04T08:29:45.551Z]       C:/Users/jenkins/workspace/_agent_elastic-agent-mbp_PR-3207/src/github.com/elastic/elastic-agent/pkg/component/runtime/command.go:422 +0x11c

[2023-10-04T08:29:45.551Z]   github.com/elastic/elastic-agent/pkg/component/runtime.(*commandRuntime).start()

[2023-10-04T08:29:45.551Z]       C:/Users/jenkins/workspace/_agent_elastic-agent-mbp_PR-3207/src/github.com/elastic/elastic-agent/pkg/component/runtime/command.go:387 +0x12de

[2023-10-04T08:29:45.551Z]   github.com/elastic/elastic-agent/pkg/component/runtime.(*commandRuntime).Run()

[2023-10-04T08:29:45.551Z]       C:/Users/jenkins/workspace/_agent_elastic-agent-mbp_PR-3207/src/github.com/elastic/elastic-agent/pkg/component/runtime/command.go:154 +0x53e

[2023-10-04T08:29:45.551Z]   github.com/elastic/elastic-agent/pkg/component/runtime.(*componentRuntimeState).runLoop.func1()

[2023-10-04T08:29:45.551Z]       C:/Users/jenkins/workspace/_agent_elastic-agent-mbp_PR-3207/src/github.com/elastic/elastic-agent/pkg/component/runtime/runtime.go:137 +0x11c

[2023-10-04T08:29:45.551Z]   github.com/elastic/elastic-agent/internal/pkg/runner.Start.func1()

[2023-10-04T08:29:45.551Z]       C:/Users/jenkins/workspace/_agent_elastic-agent-mbp_PR-3207/src/github.com/elastic/elastic-agent/internal/pkg/runner/runner.go:69 +0x62

[2023-10-04T08:29:45.551Z] 

[2023-10-04T08:29:45.551Z] Goroutine 74 (running) created at:

[2023-10-04T08:29:45.551Z]   github.com/elastic/elastic-agent/internal/pkg/runner.Start()

[2023-10-04T08:29:45.551Z]       C:/Users/jenkins/workspace/_agent_elastic-agent-mbp_PR-3207/src/github.com/elastic/elastic-agent/internal/pkg/runner/runner.go:68 +0x296

[2023-10-04T08:29:45.551Z]   github.com/elastic/elastic-agent/pkg/component/runtime.(*componentRuntimeState).runLoop()

[2023-10-04T08:29:45.551Z]       C:/Users/jenkins/workspace/_agent_elastic-agent-mbp_PR-3207/src/github.com/elastic/elastic-agent/pkg/component/runtime/runtime.go:135 +0xe8

[2023-10-04T08:29:45.551Z]   github.com/elastic/elastic-agent/pkg/component/runtime.newComponentRuntimeState.func1()

[2023-10-04T08:29:45.551Z]       C:/Users/jenkins/workspace/_agent_elastic-agent-mbp_PR-3207/src/github.com/elastic/elastic-agent/pkg/component/runtime/runtime.go:128 +0x39

[2023-10-04T08:29:45.551Z] ==================

[2023-10-04T08:29:45.551Z] ==================

[2023-10-04T08:29:45.551Z] WARNING: DATA RACE

[2023-10-04T08:29:45.551Z] Read at 0x00c0005804c8 by goroutine 75:

[2023-10-04T08:29:45.551Z]   github.com/elastic/elastic-agent/pkg/component/runtime.(*ComponentState).forceState()

[2023-10-04T08:29:45.551Z]       C:/Users/jenkins/workspace/_agent_elastic-agent-mbp_PR-3207/src/github.com/elastic/elastic-agent/pkg/component/runtime/state.go:473 +0x21c

[2023-10-04T08:29:45.551Z]   github.com/elastic/elastic-agent/pkg/component/runtime.(*commandRuntime).forceCompState()

[2023-10-04T08:29:45.551Z]       C:/Users/jenkins/workspace/_agent_elastic-agent-mbp_PR-3207/src/github.com/elastic/elastic-agent/pkg/component/runtime/command.go:311 +0x8c

[2023-10-04T08:29:45.551Z]   github.com/elastic/elastic-agent/pkg/component/runtime.(*commandRuntime).startWatcher.func1()

[2023-10-04T08:29:45.551Z]       C:/Users/jenkins/workspace/_agent_elastic-agent-mbp_PR-3207/src/github.com/elastic/elastic-agent/pkg/component/runtime/command.go:425 +0x184

[2023-10-04T08:29:45.551Z] 

[2023-10-04T08:29:45.551Z] Previous write at 0x00c0005804c8 by goroutine 74:

[2023-10-04T08:29:45.551Z]   github.com/elastic/elastic-agent/pkg/component/runtime.(*ComponentState).forceState()

[2023-10-04T08:29:45.551Z]       C:/Users/jenkins/workspace/_agent_elastic-agent-mbp_PR-3207/src/github.com/elastic/elastic-agent/pkg/component/runtime/state.go:489 +0x3a5

[2023-10-04T08:29:45.551Z]   github.com/elastic/elastic-agent/pkg/component/runtime.(*commandRuntime).forceCompState()

[2023-10-04T08:29:45.551Z]       C:/Users/jenkins/workspace/_agent_elastic-agent-mbp_PR-3207/src/github.com/elastic/elastic-agent/pkg/component/runtime/command.go:311 +0x8c

[2023-10-04T08:29:45.551Z]   github.com/elastic/elastic-agent/pkg/component/runtime.(*commandRuntime).Run()

[2023-10-04T08:29:45.551Z]       C:/Users/jenkins/workspace/_agent_elastic-agent-mbp_PR-3207/src/github.com/elastic/elastic-agent/pkg/component/runtime/command.go:160 +0x6cf

[2023-10-04T08:29:45.551Z]   github.com/elastic/elastic-agent/pkg/component/runtime.(*componentRuntimeState).runLoop.func1()

[2023-10-04T08:29:45.551Z]       C:/Users/jenkins/workspace/_agent_elastic-agent-mbp_PR-3207/src/github.com/elastic/elastic-agent/pkg/component/runtime/runtime.go:137 +0x11c

[2023-10-04T08:29:45.551Z]   github.com/elastic/elastic-agent/internal/pkg/runner.Start.func1()

[2023-10-04T08:29:45.551Z]       C:/Users/jenkins/workspace/_agent_elastic-agent-mbp_PR-3207/src/github.com/elastic/elastic-agent/internal/pkg/runner/runner.go:69 +0x62

[2023-10-04T08:29:45.551Z] 

[2023-10-04T08:29:45.551Z] Goroutine 75 (running) created at:

[2023-10-04T08:29:45.551Z]   github.com/elastic/elastic-agent/pkg/component/runtime.(*commandRuntime).startWatcher()

[2023-10-04T08:29:45.551Z]       C:/Users/jenkins/workspace/_agent_elastic-agent-mbp_PR-3207/src/github.com/elastic/elastic-agent/pkg/component/runtime/command.go:422 +0x11c

[2023-10-04T08:29:45.551Z]   github.com/elastic/elastic-agent/pkg/component/runtime.(*commandRuntime).start()

[2023-10-04T08:29:45.551Z]       C:/Users/jenkins/workspace/_agent_elastic-agent-mbp_PR-3207/src/github.com/elastic/elastic-agent/pkg/component/runtime/command.go:387 +0x12de

[2023-10-04T08:29:45.551Z]   github.com/elastic/elastic-agent/pkg/component/runtime.(*commandRuntime).Run()

[2023-10-04T08:29:45.551Z]       C:/Users/jenkins/workspace/_agent_elastic-agent-mbp_PR-3207/src/github.com/elastic/elastic-agent/pkg/component/runtime/command.go:154 +0x53e

[2023-10-04T08:29:45.551Z]   github.com/elastic/elastic-agent/pkg/component/runtime.(*componentRuntimeState).runLoop.func1()

[2023-10-04T08:29:45.551Z]       C:/Users/jenkins/workspace/_agent_elastic-agent-mbp_PR-3207/src/github.com/elastic/elastic-agent/pkg/component/runtime/runtime.go:137 +0x11c

[2023-10-04T08:29:45.551Z]   github.com/elastic/elastic-agent/internal/pkg/runner.Start.func1()

[2023-10-04T08:29:45.551Z]       C:/Users/jenkins/workspace/_agent_elastic-agent-mbp_PR-3207/src/github.com/elastic/elastic-agent/internal/pkg/runner/runner.go:69 +0x62

[2023-10-04T08:29:45.551Z] 

[2023-10-04T08:29:45.551Z] Goroutine 74 (running) created at:

[2023-10-04T08:29:45.551Z]   github.com/elastic/elastic-agent/internal/pkg/runner.Start()

[2023-10-04T08:29:45.551Z]       C:/Users/jenkins/workspace/_agent_elastic-agent-mbp_PR-3207/src/github.com/elastic/elastic-agent/internal/pkg/runner/runner.go:68 +0x296

[2023-10-04T08:29:45.552Z]   github.com/elastic/elastic-agent/pkg/component/runtime.(*componentRuntimeState).runLoop()

[2023-10-04T08:29:45.552Z]       C:/Users/jenkins/workspace/_agent_elastic-agent-mbp_PR-3207/src/github.com/elastic/elastic-agent/pkg/component/runtime/runtime.go:135 +0xe8

[2023-10-04T08:29:45.552Z]   github.com/elastic/elastic-agent/pkg/component/runtime.newComponentRuntimeState.func1()

[2023-10-04T08:29:45.552Z]       C:/Users/jenkins/workspace/_agent_elastic-agent-mbp_PR-3207/src/github.com/elastic/elastic-agent/pkg/component/runtime/runtime.go:128 +0x39

[2023-10-04T08:29:45.552Z] ==================

[2023-10-04T08:29:45.552Z]     testing.go:1446: race detected during execution of test

[2023-10-04T08:29:45.552Z] 

[2023-10-04T08:29:45.552Z] DONE 1629 tests, 19 skipped, 3 failures in 409.136s

[2023-10-04T08:29:45.552Z] Error: go test returned a non-zero value: exit status 1
```

### PR DESCRIPTION
**Pull request by AndersonQ**
## What does this PR do?

Instead of using the component state to show the error, log it using the component stderr logger

## Why is it important?

If `forceCompState` is called outside `(*commandRuntime).Run` it creates a data race.

## Checklist

- [x] My code follows the style guidelines of this project
- ~~[ ] I have commented my code, particularly in hard-to-understand areas~~
- ~~[ ] I have made corresponding changes to the documentation~~
- ~~[ ] I have made corresponding change to the default configuration files~~
- ~~[ ] I have added tests that prove my fix is effective or that my feature works~~
- ~~[ ] I have added an entry in `./changelog/fragments` using the [changelog tool](https://github.com/elastic/elastic-agent#changelog)~~
- ~~[ ] I have added an integration test or an E2E test~~

## Related issues

- Closes #3512

## Questions to ask yourself

- How are we going to support this in production? 
- How are we going to measure its adoption? 
- How are we going to debug this?
- What are the metrics I should take care of?
- ...

### CODE REVIEW

File: changelog/fragments/1696516399-Avoid-possible-data-race-when-diring-a-component-start-the-Elastic-Agent-fails-to-write-to-the-component's-stdin.yaml [hunk 1]
```diff
@@ -0,0 +1,32 @@
...
+# - enhancement: extends functionality but does not break or fix existing behavior
+# - feature: new functionality
+# - known-issue: problems that we are aware of in a given version
+# - security: impacts on the security of a product or a user’s deployment.
+# - upgrade: important information for someone upgrading from a prior version
+# - other: does not fit into any of the other categories
+kind: bug
+
+# Change summary; a 80ish characters long description of the change.
+summary: Avoid possible data race when diring a component start the Elastic Agent fails to write to the component's stdin
```
Comment (blakerouse) [inline changelog/fragments/1696516399-Avoid-possible-data-race-when-diring-a-component-start-the-Elastic-Agent-fails-to-write-to-the-component's-stdin.yaml:14]: Do we need a changelog entry? Has this actually ever effected a user in a meaningful way?

**Review by blakerouse — APPROVED** [review]
Looks good.

**Comment by AndersonQ** [comment]
buildkite test this

**Comment by jlind23** [comment]
buildkite test this

**Pull request merged**